### PR TITLE
add cycle detection for inferred outlives predicates

### LINKED ISSUE #102966: Compiler hangs on some types threading lifetimes through a carrier trait
The following code hangs rustc stable 1.64, beta and nightly. It sits on 100% CPU doing rustc_middle things inside a `rustc_typeck::outlives::implicit_infer::infer_predicates > insert_outlives_predicate` operation. (**Edit**: by "rustc_middle things" I mean `<rustc_middle::ty::TyS as core::cmp::Ord>::cmp`.) I've let it run for at least half an hour to no avail.

```rust
struct Node<'a, G: NodeGenerics<'a>> {
    kind: Kind<'a, G>,
}

enum Kind<'a, G: NodeGenerics<'a>> {
    Empty,
    Var(Var<'a, G::R>),
}

trait NodeGenerics<'a> {
    type R: 'a;
}

struct RGen<T>(std::marker::PhantomData<T>);
impl<'a, T: 'a> NodeGenerics<'a> for RGen<T> {
    type R = T;
}

struct Var<'a, R: 'a> {
    node: Box<Node<'a, RGen<R>>>,
    _phantom: std::marker::PhantomData<R>,
}
```

You can change the code in a couple of small ways such that it won't hang.

1. Remove the "carrier trait" NodeGenerics, replace it with `T: 'a` and use `T` instead of `G::R`. You would think that adding a carrier trait with the same bounds would not really change anything. But it does. (In my code, NodeGenerics has a bunch of different types attached to it, which are used in many different arms of `Kind`. It makes everything a lot neater.)
2. Changing `RGen<T>` to just `RGen` and its implementation of `NodeGenerics` to have a constant `type R = ()`.
3. Don't use lifetimes at all, put `'static` bounds in everywhere instead, i.e. make NodeGenerics take `type R: 'static`.

I don't know why it's hanging on this, but rustc should never hang, so there you go. Bug report filed.

### PR DESCRIPTION
**Pull request by cormacrelf**
This avoids a compiler hang. Fixes #102966. The solution is documented in the code, it's basically just a visit stack for the inferred predicate propagation. 

@rustbot label +A-inference +A-lifetimes +T-compiler +I-hang

### CODE REVIEW

**Review by wesleywiser — APPROVED** [review]

**Comment by wesleywiser** [comment]
This looks good to me but I'd like a second pair of eyes from

r? types

File: compiler/rustc_hir_analysis/src/outlives/implicit_infer.rs
Comment (cormacrelf) [inline compiler/rustc_hir_analysis/src/outlives/implicit_infer.rs:221]: This is the main thing I would like some eyes on. I think it's right for Var to get a substituted-through version of its own explicit bound here, because Var doesn't know what `RGen<R>::Assoc` is, but whatever it is, Node does need it to be `: 'node`. It seems the cautious approach to add the bound if we're not going to find out what the projection resolves to.

But I don't know and don't see much guidance in the RFC or its tests, as this situation wasn't really contemplated. Can anyone think of a test that might illuminate the correctness of this?

Tangentially, this PR has more tracking of the source of inferred outlives than before. Are there error messages we can improve with this information? Like the example of a "long range error" message [in the original RFC](https://rust-lang.github.io/rfcs/2093-infer-outlives.html#long-range-errors-and-why-they-are-considered-unlikely).

**Comment by compiler-errors** [comment]
Sorry, I don't really have much time to review this -- apologies for letting this sit.

r? types

**Comment by cormacrelf** [comment]
It's alright, I'm not blocked on this; the code I was working on from didn't end up benefiting from allowing a Node to have a lifetime, so I removed them. Nevertheless needs fixing at some point.

File: src/test/ui/rfc-2093-infer-outlives/issue-102966.rs
Comment (oli-obk) [inline src/test/ui/rfc-2093-infer-outlives/issue-102966.rs:14]: I don't think we should allow this code in the way you did it. This is probably unsound in case `<RGen<R> as Trait<'var>>::Assoc` does not resolve to `R`. I can check out your branch and play with a few examples to try to come up with a real bug case, but I'd prefer to error in case of a cycle and try to resolve cycles by normalizing opportunistically

File: compiler/rustc_hir_analysis/src/outlives/implicit_infer.rs
Comment (oli-obk) [inline compiler/rustc_hir_analysis/src/outlives/implicit_infer.rs:27]:
```suggestion
    'outer: for round in 1.. {
```

File: compiler/rustc_hir_analysis/src/outlives/implicit_infer.rs
Comment (oli-obk) [inline compiler/rustc_hir_analysis/src/outlives/implicit_infer.rs:101]: Just for sanity, maybe check the recursion limit here against `round` (I think that could be a valid PR on its own that we could crater and land to ensure the currently indefinite-loop actually finishes

File: src/test/ui/rfc-2093-infer-outlives/issue-102966.rs
Comment (cormacrelf) [reply src/test/ui/rfc-2093-infer-outlives/issue-102966.rs:14]: I'm happy to PR an error bailout if you want. Erroring is the conservative approach in that it does not make anything compile that previously didn't, before we're sure that's ok.

I think you'll find it's sound, though. As I understand it, inferred outlives bounds are a convenience, that relieves you from writing the bounds yourself. The only things the cycle detector can do wrong are:

- Kick in "too early", i.e. not adding enough bounds. Basically, if you fail to add the necessary bounds, *rustc is still checking your structs internally for whether the bounds of the types you use are met*.  The Rust reference type `&'a T` has a bound on it relating the referent type and the lifetime. You have to prove this any time you use it, and adding a `T: 'a` bound to `struct A<'a, T>(&'a T)` is just kicking this down the road to whoever uses `A`. Before RFC2093, the kicking of that can down the road was not automatic, and so writing that struct definition would result in the `&'a T` type complaining its bounds weren't met, and you having to writing some manual bounds. That's the worst case scenario.
- Kick in "too late", i.e. add too many bounds. This is definitionally sound. You can never produce unsoundness by adding more constraints on the lifetime of a type parameter. Maybe the resulting type is less versatile than it could be, but it won't be unsound.

For posterity, I did think about the whole "what if `RGen<R>::Assoc` is not `R` thing, and it is a red herring. The equivalence is not how the cycle was formed: you can change it to

```rust
trait Trait<'a> { type Assoc; /* no bound */ }
impl<'a, R> Trait<'a> for RGen<R> { type Assoc = String }
```

... and there is still a cycle and rustc hangs, because the cycle comes from the repeated substitution of `Var`'s `R: 'var` through projections that create more names for the same types. The solution to the "what outlives bounds do we need to add to satisfy the types inside `Node` and `Var`" problem is just "`Node` gets `T::Assoc: 'node`". Don't trust me, trust a pre-RFC2093 compiler (1.12.1):

```
error[E0309]: the associated type `<T as issue::Trait<'node>>::Assoc` may not live long enough
 --> src/issue.rs:5:5
  |
5 |     var: Var<'node, T::Assoc>,
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^
  |
  = help: consider adding an explicit lifetime bound `<T as issue::Trait<'node>>::Assoc: 'node`...
note: ...so that the type `<T as issue::Trait<'node>>::Assoc` will meet its required lifetime bounds
 --> src/issue.rs:5:5
  |
5 |     var: Var<'node, T::Assoc>,
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^
```

When you add that bound, rustc-1.12.1 is happy. It turns out that only one additional bound was necessary to infer here, so my propagation solution exhibits the problem of "kicking in too late". Perhaps I should revisit the choice not to add the site of the original explicit bound as part of the provenance.

Comment (cormacrelf) [reply src/test/ui/rfc-2093-infer-outlives/issue-102966.rs:14]: Ohhhh, this is how it can be unsound: https://github.com/rust-lang/rust/issues/84305. I get you now. I still think this will turn out sound but yeah, I will PR a simple recursion limit first.

**Comment by lcnr** [comment]
waiting and author and doesn't seem to need t-types input rn

https://rust-lang.zulipchat.com/#narrow/stream/326132-t-types.2Fmeetings/topic/2023-01-18.20meeting/near/322081814

**Comment by JohnCSimon** [comment]
@cormacrelf 

ping from triage - can you post your status on this PR? There hasn't been an update in a few months. Thanks!


FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

**Comment by Dylan-DPC** [comment]
@cormacrelf any updates on this?

**Comment by cormacrelf** [comment]
Hi, sorry I haven't been working on this. I started to rework this into a more conservative "error with recursion limit exceeded" as requested, but I didn't finish it as it turned out to be a pretty complicated dance to get a new kind of compiler message going. There is just a lot of error messaging API and it is very decoupled.

I haven't been pressed to finish it because the library I wanted this kind of lifetime arrangement for didn't end up needing it: <https://github.com/cormacrelf/incremental-rs>. But I will get there. I think I have a few commits that are halfway there that I can put up at least.

**Comment by JohnCSimon** [comment]
@cormacrelf 
Ping from triage: I'm closing this due to inactivity, Please reopen when you are ready to continue with this. 
Note: if you are going to continue please open the PR **BEFORE** you push to it, else you won't be able to reopen - this is a quirk of github. 
Thanks for your contribution.

@rustbot label: +S-inactive

**Pull request closed without merging**
